How-To: Increase Storage Size on an OCP Logging Elasticsearch instance (#151)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/logging/increase-elasticsearch-storage-size.adoc

= Increase Storage Size of an OCP Logging Elasticsearch instance

[abstract]
This page describes how to increase the underlying storage size of the OpenShift Cluster Logging Elasticsearch instance.

> 

Reviewer: I think we should mention the impact of the resize.
```suggestion

[IMPORTANT]
This is a disruptive operation!
During the resize, the Elasticsearch cluster will experience downtime for a few minutes to hours.

```

Author: There is no downtime since we have single redundancy!

Author: There will be reduced performance for a few minutes though and new logs are delayed. I try to write it like this.

== Prerequisites

* `kubectl`
* `curl`
* `jq`
* `yq` https://mikefarah.gitbook.io/yq[yq YAML processor] (version 4 or higher)
* `commodore`, see https://syn.tools/commodore/running-commodore.html[Running Commodore]

== Prepare local environment

. Configure API access
+
[source,bash]
----
export COMMODORE_API_URL=https://api.syn.vshn.net <1>
export COMMODORE_API_TOKEN=<your lieutenant token>

# Set Project Syn cluster and tenant ID
export CLUSTER_ID=<lieutenant-cluster-id> # Looks like: c-cluster-id-1234
export TENANT_ID=$(curl -sH "Authorization: Bearer ${COMMODORE_API_TOKEN}" ${COMMODORE_API_URL}/clusters/${CLUSTER_ID} | jq -r .tenant)
----
<1> Replace with the API URL of the desired Lieutenant instance.

. Create a local directory to work in and compile the cluster catalog
+
[source,bash]
----
export WORK_DIR=/path/to/work/dir
mkdir -p "${WORK_DIR}"
pushd "${WORK_DIR}"

commodore catalog compile "${CLUSTER_ID}"
----
+
[TIP]
====
We strongly recommend creating an empty directory, unless you already have a work directory for the cluster you're about to work on.
This guide will run Commodore in the directory created in this step.

> ====

Reviewer: I think here we can just link to https://syn.tools/syn/how-tos/compile-catalog.html

Author: I'm not a fan. In my experience it's way easier to follow a guide on one page and not have to jump between 2-3 pages.

It would be a bit inconsistent too. All other mentions in the docs show how to compile commodore directly on the page and don't link to syn.tools.

Reviewer: Ok, but at least make a partial, otherwise this is unmaintainable.

Author: 👍

== Increase PVC sizes

=== Update Catalog

. Set desired size
+
[source,bash]
----
STORAGE_SIZE=250Gi <1>
----
<1> Replace with the desired PVC size.

. Update Commodore catalog
+
[source,bash]
----
pushd "inventory/classes/${TENANT_ID}/"

yq eval -i "parameters.openshift4_logging.clusterLogging.logStore.elasticsearch.storage.size += \"${STORAGE_SIZE}\"" \
  ${CLUSTER_ID}.yml

git commit -a -m "Set Elasticsearch backing storage to \"${STORAGE_SIZE}\" on ${CLUSTER_ID}"
git push
popd
----
+
. Compile catalog
+
[source,bash]
----
commodore catalog compile ${CLUSTER_ID} --push -i

> ----

Reviewer: Maybe this can also be combined with https://syn.tools/syn/how-tos/change-parameter.html

+
[source,bash]
----
kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  patch daemonset fluentd \
  -p '{"spec":{"template":{"spec":{"nodeSelector":{"logging-infra-fluentd": "false"}}}}}'
----
+
. Perform a shard synced flush using the `es_util` tool to ensure there are no pending operations waiting to be written to disk prior to shutting down.
+
[source,bash]
----
es_pod=$(kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  get pods \
  -l component=elasticsearch \
  -o name | head -n1)

kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  exec "${es_pod}" \
  -c elasticsearch \
  -- es_util --query="_flush/synced" -XPOST
----
+
Example output:
+
[source,json]
----
{"_shards":{"total":4,"successful":4,"failed":0},".security":{"total":2,"successful":2,"failed":0},".kibana_1":{"total":2,"successful":2,"failed":0}}
----
+
. Prevent shard balancing when purposely bringing down nodes.
+
[source,bash]
----
kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  exec "${es_pods[1]}" \
  -c elasticsearch \
  -- es_util --query="_cluster/settings" -XPUT -d '{ "persistent": { "cluster.routing.allocation.enable" : "primaries" } }'
----
+

> Example output:
> +
> [source,json]

Reviewer: Maybe just give the box a title? (applies to other locations as well)
```suggestion
.Example output
[source,json]
```

----
+
. Find Elasticsearch deployments
+
[source,bash]
----
kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  get deploy \
  -l component=elasticsearch
----
+
Sample output:
+
[source]
----
NAME                           READY   UP-TO-DATE   AVAILABLE   AGE
elasticsearch-cdm-7ya69va8-1   1/1     1            1           68d
elasticsearch-cdm-7ya69va8-2   1/1     1            1           68d
elasticsearch-cdm-7ya69va8-3   1/1     1            1           68d
----
+
. For each deployment do
.. Restart Elasticsearch
+
[source,bash]
----
ES_DEPLOYMENT=elasticsearch-cdm-7ya69va8-1 <1>

kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  scale deploy/${ES_DEPLOYMENT} \
  --replicas=0

# Verify pod is removed
kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  get pods \
  | grep "${ES_DEPLOYMENT}-"

kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  scale deploy/${ES_DEPLOYMENT} \
  --replicas=1

# Wait for pod to become ready
kubectl \
  --as=cluster-admin \
  -n openshift-logging \
  get pods \
  --watch
----
<1> Replace with deployment name found in previous step.
+
.. Wait until cluster becomes healthy again.
+
[WARNING]

> Make sure the status is `green` or `yellow` before proceeding.

Reviewer: I think we should decide on which "color", this is unclear
```suggestion
Make sure the status is at least `yellow` before proceeding.
```

Author: I thing it's clearer if both possible colors are explicitly mentioned. For me the cluster stayed green while restarting two of three nodes.
